tracer/src/Datadog.InstrumentedAssemblyGenerator/ModifiedMethod.cs

```diff
@@ -0,0 +1,158 @@
+            _methodNameIndex = FullName.Substring(0, FullName.LastIndexOf('(')).LastIndexOf(".", StringComparison.InvariantCulture);
+            DecompiledCode = new Lazy<string>(Decompile);
+            IsValid = true;
+        }
+        catch (Exception e)
+        {
+            IsValid = false;
+            Console.WriteLine(e);
+            ModulePath = null;
+            Instructions = null;
+            Arguments = null;
+            FullName = null;
+            _methodNameIndex = -1;
+        }
+    }
+
+    /// <summary>
+    /// If the current instance is a valid ModifiedMethod object
+    /// </summary>
+    public bool IsValid { get; set; }
+
+    /// <summary>
+    /// The module of the method
+    /// </summary>
+    public string ModulePath { get; }
+
+    /// <summary>
+    /// the full name of the method
+    /// </summary>
+    public string FullName { get; }
+
+    /// <summary>
+    /// The name of the method parent type
+    /// </summary>
+    public string TypeName
+    {
+        get
+        {
+            if (!IsValid)
+            {
+                return null;
+            }
+
+            if (MethodName.StartsWith("ctor") || MethodName.StartsWith("cctor"))
+            {
+                return FullName.Substring(0, _methodNameIndex - 1);
+            }
+            else
+            {
+                return FullName.Substring(0, _methodNameIndex);
+            }
+        }
+    }
+
+    /// <summary>
+    /// The name of the method including method arguments
+    /// </summary>
+    public string MethodName
+    {
+        get
+        {
+            if (!IsValid)
+            {
+                return null;
+            }
+
+            return FullName.Substring(_methodNameIndex + 1);
+        }
+    }
+
+    /// <summary>
+    /// The full name arguments of the method
+    /// </summary>
+    public List<string> Arguments { get; }
+
+    /// <summary>
+    /// The IL instructions of the method
+    /// </summary>
+    public string Instructions { get; }
+
+    /// <summary>
+    /// Get the decompiled code of the method
+    /// </summary>
+    public Lazy<string> DecompiledCode { get; }
+
+    private string Decompile()
+    {
+        if (!IsValid)
+        {
+            return null;
+        }
+
+        try
+        {
+            var settings = new DecompilerSettings();
+            var decompiler = new CSharpDecompiler(ModulePath, settings);
+            var typeName = new FullTypeName(TypeName);
+            ITypeDefinition typeInfo = decompiler.TypeSystem.MainModule.Compilation.FindType(typeName).GetDefinition();
+            if (typeInfo == null)
+            {
+                Console.WriteLine($"{nameof(ModifiedMethod.Decompile)}: Can't find type {TypeName}");
+                return null;
+            }
+
+            // we can't do this replace in ILSpy method because '`' is valid sign for generic type
+            var copyOfMethodAndParameterName = MethodName.Replace("!", "`");
```

Comment:
This seems quite error prone. Couldn't we reply solely on the method tokens that were generated for the instrumented methods?

Reply:
I agree, the problem that we handle with two type systems - dnlib and ILSpy. Is it possible to resolve a regular method based on token. But to resolve a generic method we must have in hand the generic context, which we have in dnlib form, and to create the correct generic context for ILSpy, we must have the owner of that context, which is the method that we search for.

Reply:
See 869b8f9aad16a4ab73809f24951be5c938bb7e2f